chain/network/src/routing/actor.rs

```diff
@@ -0,0 +1,319 @@
+    ///
+    /// Step 3) An active edge gets added from `D` to `A`.
+    /// We will load `C_1` and try to re-add all edges belonging to `C_1`.
+    /// We will add `<A,B>, <A,C>, <A, D>, <B, C>, <B, D>`
+    ///
+    /// Important note: `C_1` also contains an edge from `A` to `C`, though `C` was removed in `C_2`.
+    /// - 1) We will not load edges belonging to `C_2`, even though we are adding an edges from `A` to deleted `C`.
+    /// - 2) We will not delete mapping from `C` to `C_2`, because `C` doesn't belong to `C_1`.
+    /// - 3) Later, `C` will be deleted, because we will figure out it's not reachable.
+    /// New component `C_3` will be created.
+    /// And mapping from `C` to `C_2` will be overridden by mapping from `C` to `C_3`.
+    /// And therefore `C_2` component will become unreachable.
+    /// TODO: this whole algorithm seems to be leaking stuff to storage and never cleaning up.
+    /// What is the point of it? What does it actually gives us?
+    fn load_component(&mut self, peer_id: &PeerId) {
+        if *peer_id == self.my_peer_id || self.peer_reachable_at.contains_key(peer_id) {
+            return;
+        }
+        let edges = match self.store.pop_component(peer_id) {
+            Ok(edges) => edges,
+            Err(e) => {
+                warn!("self.store.pop_component({}): {}", peer_id, e);
+                return;
+            }
+        };
+        self.graph.write().update_edges(edges);
+    }
+
+    /// Prunes peers unreachable since <unreachable_since> (and their adjacent edges)
+    /// from the in-mem graph and stores them in DB.
+    /// # Returns
+    /// List of edges removed.
+    pub(crate) fn prune_unreachable_peers(
+        &mut self,
+        unreachable_since: time::Instant,
+    ) -> Vec<Edge> {
+        let _d = delay_detector::DelayDetector::new(|| "pruning unreachable peers".into());
+
+        // Select peers to prune.
+        let mut peers = HashSet::new();
+        for (k, _) in self.graph.read().edges() {
+            for peer_id in [&k.0, &k.1] {
+                if self
+                    .peer_reachable_at
+                    .get(peer_id)
+                    .map(|t| t < &unreachable_since)
+                    .unwrap_or(true)
+                {
+                    peers.insert(peer_id.clone());
+                }
+            }
+        }
+        if peers.is_empty() {
+            return vec![];
+        }
+
+        // Prune peers from peer_reachable_at.
+        for peer_id in &peers {
+            self.peer_reachable_at.remove(&peer_id);
+        }
+
+        // Prune edges from graph.
+        let edges = self.graph.write().remove_adjacent_edges(&peers);
+
+        // Store the pruned data in DB.
+        if let Err(e) = self.store.push_component(&peers, &edges) {
+            warn!("self.store.push_component(): {}", e);
+        }
+        edges
+    }
+
+    pub fn update_routing_table(
+        &mut self,
+        mut prune_unreachable_since: Option<time::Instant>,
```

Comment:
does it have to be mut ?

Reply:
Yes, because it updates peer_reachable_at map.